File: web/js/advisor.js
/**
 * Web 版决策助手：用户输入当前局面 → 调 ExpectimaxAgent 给建议 + 概率分解。
 * 镜像 advisor.py 的逻辑。
 */

import {
  BONUS_FLAT_AVG, BONUS_FLAT_VALUES, DeckCounts,
} from './cards.js';
import {
  GameConfig, GameState, PlayerState, PlayerStatus,
} from './state.js';
import { ExpectimaxAgent } from './agents.js';

// ============ 状态 ============
const myHand = []; // [3, 5, 8] 手牌数字（可能重复出现，但 unique 才算手牌）
const seenNums = {}; // {3: 2, 5: 1, ...} 已见过的数字（含其他玩家手牌+弃牌堆）
const seenSpecials = { bonus_flat: 0, bonus_double: 0, insurance: 0, exile: 0, triple: 0 };

// ============ 初始化按钮 ============
function buildPickers() {
  // 我的手牌：13 个数字，点击切换是否在手中
  const myDiv = document.getElementById('my-hand-picker');
  for (let n = 0; n <= 12; n++) {
    const btn = document.createElement('div');
    btn.className = 'pick-btn';
    btn.textContent = n;
    btn.dataset.num = n;
    btn.addEventListener('click', () => toggleMyHand(n));
    myDiv.appendChild(btn);
  }

  // 已见数字：点 +1，右键 / 长按 -1
  const numDiv = document.getElementById('seen-num-picker');
  for (let n = 0; n <= 12; n++) {
    const btn = makeCounter('seen-num-' + n, n.toString(), () => addSeenNum(n), () => subSeenNum(n));
    btn.dataset.num = n;
    numDiv.appendChild(btn);
  }

  // 已见加分牌
  const bonusDiv = document.getElementById('seen-bonus-picker');
  for (const v of BONUS_FLAT_VALUES) {
    const btn = makeCounter(`seen-flat-${v}`, `+${v}`,
      () => addSeenSpecial('bonus_flat'), () => subSeenSpecial('bonus_flat'));
    bonusDiv.appendChild(btn);
  }
  bonusDiv.appendChild(makeCounter('seen-double', '×2',
    () => addSeenSpecial('bonus_double'), () => subSeenSpecial('bonus_double')));

  // 已见技能牌
  const skillDiv = document.getElementById('seen-skill-picker');
  skillDiv.appendChild(makeCounter('seen-ins', '🛡 保险',
    () => addSeenSpecial('insurance'), () => subSeenSpecial('insurance')));
  skillDiv.appendChild(makeCounter('seen-exile', '🚷 放逐',
    () => addSeenSpecial('exile'), () => subSeenSpecial('exile')));
  skillDiv.appendChild(makeCounter('seen-triple', '⚡ 三连',
    () => addSeenSpecial('triple'), () => subSeenSpecial('triple')));
}

function makeCounter(id, label, onAdd, onSub) {
  const btn = document.createElement('div');
  btn.className = 'pick-btn';
  btn.id = id;
  btn.innerHTML = `<span>${label}</span><span class="count" data-count="0"></span>`;
  btn.addEventListener('click', onAdd);
  btn.addEventListener('contextmenu', e => { e.preventDefault(); onSub(); });
  // 长按 -1（移动端）
  let timer = null;
  btn.addEventListener('touchstart', () => {
    timer = setTimeout(() => { onSub(); timer = null; }, 500);
  });
  btn.addEventListener('touchend', () => {
    if (timer) clearTimeout(timer);
  });
  return btn;
}

function toggleMyHand(n) {
  const idx = myHand.indexOf(n);
  if (idx >= 0) myHand.splice(idx, 1);
  else myHand.push(n);
  refreshMyHand();
}

function refreshMyHand() {
  document.querySelectorAll('#my-hand-picker .pick-btn').forEach(btn => {
    const n = parseInt(btn.dataset.num);
    btn.classList.toggle('active', myHand.includes(n));
  });
  const display = document.getElementById('my-hand-display');
  if (myHand.length === 0) {
    display.textContent = '点下面数字加入手牌：';
  } else {
    const sorted = myHand.slice().sort((a, b) => a - b);
    display.innerHTML = `当前手牌: <b style="color:var(--accent);font-family:ui-monospace,monospace">[${sorted.join(', ')}]</b>`;
  }
}

function addSeenNum(n) {
  seenNums[n] = (seenNums[n] || 0) + 1;
  if (seenNums[n] > (n === 0 || n === 1 ? 1 : n)) {
    seenNums[n] = (n === 0 || n === 1 ? 1 : n);  // cap to deck
  }
  refreshSeenNum(n);
}
function subSeenNum(n) {
  if (seenNums[n] > 0) seenNums[n] -= 1;
  refreshSeenNum(n);
}
function refreshSeenNum(n) {
  const btn = document.getElementById('seen-num-' + n);
  const c = seenNums[n] || 0;
  btn.classList.toggle('active', c > 0);
  btn.querySelector('.count').textContent = c > 0 ? c : '';
}

function addSeenSpecial(key) {
  seenSpecials[key] = (seenSpecials[key] || 0) + 1;
  // cap
  const max = { bonus_flat: 5, bonus_double: 3, insurance: 3, exile: 3, triple: 3 }[key];
  if (seenSpecials[key] > max) seenSpecials[key] = max;
  refreshSeenSpecial(key);
}
function subSeenSpecial(key) {
  if (seenSpecials[key] > 0) seenSpecials[key] -= 1;
  refreshSeenSpecial(key);
}
function refreshSeenSpecial(key) {
  const ids = {
    bonus_flat: BONUS_FLAT_VALUES.map(v => 'seen-flat-' + v),
    bonus_double: ['seen-double'],
    insurance: ['seen-ins'],
    exile: ['seen-exile'],
    triple: ['seen-triple'],
  };
  // 加分牌 5 张面值，count 平均显示在每个按钮上
  ids[key].forEach((id, i) => {
    const btn = document.getElementById(id);
    if (!btn) return;
    const total = seenSpecials[key] || 0;
    btn.classList.toggle('active', total > 0);
    if (key === 'bonus_flat') {
      // 不知道用户具体见过哪个面值，全部按 active 显示，count 显示总数在第一个
      btn.querySelector('.count').textContent = (i === 0 && total > 0) ? total : '';
    } else {
      btn.querySelector('.count').textContent = total > 0 ? total : '';
    }
  });
}

// ============ 构建 state ============
function buildRemaining() {
  const rem = DeckCounts.full();
  // 减自己手牌的数字
  for (const v of myHand) rem.numbers[v] -= 1;
  // 减自己的加分（按数量贪心推算）
  const myBonus = parseInt(document.getElementById('my-bonus').value) || 0;
  let needed = myBonus;
  let myBonusFlatCount = 0;
  for (const v of [10, 8, 6, 4, 2]) {
    if (needed >= v) { myBonusFlatCount += 1; needed -= v; }
  }
  rem.bonus_flat -= myBonusFlatCount;
  if (document.getElementById('my-double').checked) rem.bonus_double -= 1;
  if (document.getElementById('my-insurance').checked) rem.insurance -= 1;
  // 减已见
  for (const [v, c] of Object.entries(seenNums)) rem.numbers[v] -= c;
  rem.bonus_flat -= seenSpecials.bonus_flat;
  rem.bonus_double -= seenSpecials.bonus_double;
  rem.insurance -= seenSpecials.insurance;
  rem.exile -= seenSpecials.exile;
  rem.triple -= seenSpecials.triple;

  // 防负
  const issues = [];
  for (let v = 0; v <= 12; v++) {
    if (rem.numbers[v] < 0) {
      issues.push(`数字 ${v} 多算了 ${-rem.numbers[v]} 张`);
      rem.numbers[v] = 0;
    }
  }
  for (const k of ['bonus_flat', 'bonus_double', 'insurance', 'exile', 'triple']) {
    if (rem[k] < 0) {
      issues.push(`${k} 多算了 ${-rem[k]} 张`);
      rem[k] = 0;
    }
  }
  return { rem, issues, myBonusFlatCount };
}

// ============ 分析 ============
function analyze() {
  const { rem, issues } = buildRemaining();
  const myBonus = parseInt(document.getElementById('my-bonus').value) || 0;
  const myTotal = parseInt(document.getElementById('my-total').value) || 0;
  const target = parseInt(document.getElementById('target').value) || 200;
  const numPlayers = parseInt(document.getElementById('num-players').value) || 4;
  const hasInsurance = document.getElementById('my-insurance').checked;

  // 构 state（其他玩家粗暴当 active 总分 0；advisor 主要看自己决策）
  const cfg = new GameConfig({ numPlayers, targetScore: target });
  const me = new PlayerState(0);
  me.totalScore = myTotal;
  me.handNumbers = myHand.slice();
  me.bonusFlatTotal = myBonus;
  me.hasInsurance = hasInsurance;
  me.status = PlayerStatus.ACTIVE;
  const others = [];
  for (let i = 1; i < numPlayers; i++) {
    const p = new PlayerState(i);
    p.status = PlayerStatus.ACTIVE;
    others.push(p);
  }
  const state = new GameState(cfg, [me, ...others], rem);

  // 概率分解
  const total = Math.max(rem.total(), 1);
  const handSet = me.uniqueNumbers;
  const nUnique = handSet.size;
  const curScore = myHand.reduce((s, v) => s + v, 0) + myBonus;

  let bustCount = 0;
  for (const v of handSet) bustCount += rem.numbers[v];
  let safeCount = 0;
  let safeSum = 0;
  for (let v = 0; v <= 12; v++) {
    if (!handSet.has(v)) { safeCount += rem.numbers[v]; safeSum += v * rem.numbers[v]; }
  }
  const pBust = bustCount / total;
  const pSafe = safeCount / total;
  const pSix = nUnique === 5 ? pSafe : 0;
  const pContSafe = nUnique < 5 ? pSafe : 0;
  const avgSafeValue = safeCount > 0 ? safeSum / safeCount : 0;

  const pFlat = rem.bonus_flat / total;
  const pDouble = rem.bonus_double / total;
  const pIns = rem.insurance / total;
  const pExile = rem.exile / total;
  const pTriple = rem.triple / total;

  // EV 估算
  const evFold = curScore;
  let evDraw = 0;
  evDraw += pBust * (hasInsurance ? curScore : 0);
  evDraw += pContSafe * (curScore + avgSafeValue);
  evDraw += pSix * (curScore + avgSafeValue + 15);
  evDraw += pFlat * (curScore + BONUS_FLAT_AVG);
  evDraw += pDouble * (curScore + myHand.reduce((s, v) => s + v, 0));
  evDraw += (pIns + pExile + pTriple) * (curScore + 5);

  // 调真正的 expectimax
  const expectimax = new ExpectimaxAgent(3);
  let exDecision;
  try {
    exDecision = expectimax.chooseAction(state, 0);
  } catch (e) {
    console.error(e);
    exDecision = evDraw > evFold ? 'draw' : 'fold';
  }

  return {
    issues, totalRemaining: total, nUnique, curScore, hasInsurance,
    myTotal, target,
    pBust: pBust * 100, pSafe: pSafe * 100, pSix: pSix * 100, avgSafeValue,
    pFlat: pFlat * 100, pDouble: pDouble * 100,
    pIns: pIns * 100, pExile: pExile * 100, pTriple: pTriple * 100,
    evDraw, evFold, exDecision,
    handSum: myHand.reduce((s, v) => s + v, 0),
  };
}

// ============ 渲染结果 ============
function render(a) {
  document.getElementById('result').hidden = false;
  // 状态摘要
  const summaryDiv = document.getElementById('state-summary');
  const handStr = myHand.length ? myHand.slice().sort((x,y)=>x-y).join(', ') : '空';
  summaryDiv.innerHTML = `
    <div>剩余牌库 <b>${a.totalRemaining}</b> 张</div>
    <div>手牌 <b>[${handStr}]</b> · 加分 <b>+${a.curScore - a.handSum}</b> · 不同数字 <b>${a.nUnique}</b> 个</div>
    <div>本局已得 <b>${a.curScore}</b> · 总分 <b>${a.myTotal} / ${a.target}</b> · 保险 <b>${a.hasInsurance ? '有 🛡' : '无'}</b></div>
  `;

  // banners
  const bannerDiv = document.getElementById('banners');
  bannerDiv.innerHTML = '';
  if (a.issues.length > 0) {
    bannerDiv.innerHTML += `<div class="warning-banner">⚠ 输入与牌库矛盾：${a.issues.join('；')}</div>`;
  }
  if (a.myTotal + a.curScore >= a.target) {
    bannerDiv.innerHTML += `<div class="winning-banner">🏆 现在跑路 锁 ${a.curScore} 分即可获胜！</div>`;
  }
  if (a.pBust > 35) {
    bannerDiv.innerHTML += `<div class="warning-banner">⚠ 高风险：爆牌概率 ${a.pBust.toFixed(0)}%，强烈建议跑路</div>`;
  }

  // 概率分解
  const probDiv = document.getElementById('prob-detail');
  probDiv.innerHTML = '';
  const cls = (p, low, mid) => p < low ? 'safe' : p < mid ? 'warn' : 'danger';
  const rows = [];
  rows.push({
    emoji: '💥', label: '爆牌', pct: a.pBust, cls: cls(a.pBust, 15, 30),
    note: a.hasInsurance ? '有保险，免一次' : `本局 ${a.curScore} 分归零`,
  });
  if (a.nUnique < 5) {
    rows.push({
      emoji: '✅', label: '安全数字', pct: a.pSafe, cls: 'safe',
      note: `平均 +${a.avgSafeValue.toFixed(1)} 分`,
    });
  } else if (a.nUnique === 5) {
    rows.push({
      emoji: '🚀', label: '6 翻终结', pct: a.pSix, cls: a.pSix > 50 ? 'safe' : 'warn',
      note: '+15 奖励 + 全场强制结算',
    });
  }
  rows.push({ emoji: '🟢', label: '加分牌（平均 +6）', pct: a.pFlat, cls: 'safe', note: '+2/+4/+6/+8/+10' });
  rows.push({ emoji: '🟢', label: '翻倍', pct: a.pDouble, cls: 'safe', note: `+${a.handSum} 分` });
  rows.push({ emoji: '🛡', label: '保险', pct: a.pIns, cls: 'safe', note: '免一次爆牌' });
  rows.push({ emoji: '🚷', label: '放逐', pct: a.pExile, cls: 'safe', note: '强制目标跑路' });
  rows.push({ emoji: '⚡', label: '三连', pct: a.pTriple, cls: 'safe', note: '强制连摸 3 张' });

  rows.forEach(r => {
    const el = document.createElement('div');
    el.className = 'prob-row ' + r.cls;
    el.innerHTML = `
      <span>${r.emoji}</span>
      <span>${r.label}</span>
      <span class="pct">${r.pct.toFixed(1)}%</span>
      <span class="note">${r.note}</span>
    `;
    probDiv.appendChild(el);
  });

  // EV 对比
  const evDiv = document.getElementById('ev-compare');
  const drawWins = a.evDraw > a.evFold;
  evDiv.innerHTML = `
    <div class="ev-card ${drawWins ? 'win' : 'lose'}">
      <div class="label">EV(继续摸)</div>
      <div class="value ${drawWins ? 'win' : ''}">${a.evDraw.toFixed(1)}</div>
    </div>
    <div class="ev-card ${!drawWins ? 'win' : 'lose'}">
      <div class="label">EV(跑路)</div>
      <div class="value ${!drawWins ? 'win' : ''}">${a.evFold.toFixed(1)}</div>
    </div>
  `;

  // 推荐
  const recDiv = document.getElementById('recommend');
  if (a.exDecision === 'draw') {
    recDiv.innerHTML = `
      <div class="recommend">
        <div class="verdict">🎲 继续摸牌</div>
        <div class="reason">Expectimax depth=3 计算后建议</div>
      </div>
    `;
  } else {
    recDiv.innerHTML = `
      <div class="recommend fold">
        <div class="verdict">🔒 跑路锁分</div>
        <div class="reason">Expectimax depth=3 计算后建议（锁定 ${a.curScore} 分）</div>
      </div>
    `;
  }
}

// ============ 入口 ============
document.addEventListener('DOMContentLoaded', () => {
  buildPickers();
  refreshMyHand();
  document.getElementById('analyze-btn').addEventListener('click', () => {
    try {
      const a = analyze();
      render(a);
      document.getElementById('result').scrollIntoView({ behavior: 'smooth', block: 'start' });
    } catch (e) {
      console.error(e);
      alert('分析失败：' + (e.message || e));
    }
  });
});

File: advisor.py
"""决策助手：输入当前局面，给出 draw / fold 建议 + 详细概率分析。

适用场景：你在线下/别的平台玩这个游戏，用这个工具实时辅助决策。

用法：
    python3 advisor.py
    python3 advisor.py --hand "3 5 8" --bonus 10 --total 45 --target 200
"""

from __future__ import annotations

import argparse
import os
import sys
from typing import Dict, List, Optional, Tuple

from agents.ev_agent import EVAgent
from agents.expectimax_agent import ExpectimaxAgent
from game import GameConfig
from game.cards import (BONUS_FLAT_COUNT, BONUS_DOUBLE_COUNT, NUMBER_COUNTS,
                        SKILL_PER_KIND, DeckCounts)
from game.state import GameState, PlayerState, PlayerStatus


USE_COLOR = sys.stdout.isatty() and os.environ.get("TERM") != "dumb"
def C(c, t):
    return f"\033[{c}m{t}\033[0m" if USE_COLOR else t
BOLD = lambda t: C("1", t)
DIM = lambda t: C("2", t)
GREEN = lambda t: C("32", t)
YELLOW = lambda t: C("33", t)
RED = lambda t: C("31", t)
CYAN = lambda t: C("36", t)
MAGENTA = lambda t: C("35", t)


# 中文别名 → 内部名 (+加分值)
# 加分牌：+2/+4/+6/+8/+10 各 1 张；翻倍牌单独 3 张
SPECIAL_ALIASES = {
    # +N 加分牌（具体面值，但 DeckCounts 只记总张数，所以只用 'bonus_flat'）
    "+2": "bonus_flat", "+4": "bonus_flat", "+6": "bonus_flat",
    "+8": "bonus_flat", "+10": "bonus_flat",
    "加分": "bonus_flat", "flat": "bonus_flat",
    "翻倍": "bonus_double", "double": "bonus_double", "x2": "bonus_double", "×2": "bonus_double",
    "保险": "insurance", "ins": "insurance", "shield": "insurance",
    "放逐": "exile", "exile": "exile",
    "三连": "triple", "triple": "triple",
}


def parse_seen(s: str) -> Tuple[Dict[int, int], Dict[str, int]]:
    """解析 '1 2 7 +10 保险' → ({1:1, 2:1, 7:1}, {bonus_flat:1, insurance:1})"""
    nums: Dict[int, int] = {}
    specials: Dict[str, int] = {}
    if not s.strip():
        return nums, specials
    for tok in s.replace(",", " ").split():
        tok = tok.strip()
        if not tok:
            continue
        if tok.lstrip("-").isdigit() and not tok.startswith("+"):
            v = int(tok)
            if 0 <= v <= 12:
                nums[v] = nums.get(v, 0) + 1
                continue
        key = SPECIAL_ALIASES.get(tok.lower())
        if key is None:
            print(RED(f"  ⚠ 未识别的牌：'{tok}'，跳过"))
            continue
        specials[key] = specials.get(key, 0) + 1
    return nums, specials


def parse_hand(s: str) -> List[int]:
    out = []
    for tok in s.replace(",", " ").split():
        try:
            v = int(tok)
            if 0 <= v <= 12:
                out.append(v)
        except ValueError:
            pass
    return out


def build_remaining(my_hand: List[int], my_bonus_flat_count: int,
                     my_bonus_double_count: int, my_insurance: bool,
                     seen_nums: Dict[int, int], seen_specials: Dict[str, int]) -> DeckCounts:
    rem = DeckCounts.full()
    # 减去自己手上的数字
    for v in my_hand:
        rem.numbers[v] -= 1
    rem.bonus_flat -= my_bonus_flat_count
    rem.bonus_double -= my_bonus_double_count
    if my_insurance:
        rem.insurance -= 1
    # 减去已见
    for v, c in seen_nums.items():
        rem.numbers[v] -= c
    for k, c in seen_specials.items():
        setattr(rem, k, getattr(rem, k) - c)
    # 防负值并提示
    issues = []
    for v in range(13):
        if rem.numbers[v] < 0:
            issues.append(f"数字{v}多算了 {-rem.numbers[v]} 张")
            rem.numbers[v] = 0
    for k in ("bonus_flat", "bonus_double", "insurance", "exile", "triple"):
        if getattr(rem, k) < 0:
            issues.append(f"{k} 多算了 {-getattr(rem, k)} 张")
            setattr(rem, k, 0)
    if issues:
        print(RED("  ⚠ 输入与牌库矛盾：" + "；".join(issues)))
    return rem


def analyze(state: GameState, my_idx: int) -> dict:
    me = state.players[my_idx]
    counts = state.remaining
    total = counts.total()
    cur_score = sum(me.hand_numbers) + me.bonus_flat_total
    hand_set = me.unique_numbers
    n_unique = len(hand_set)

    if total == 0:
        return {"error": "牌库已空"}

    bust_count = sum(counts.numbers[v] for v in hand_set)
    safe_count = sum(c for v, c in counts.numbers.items() if v not in hand_set)
    p_bust = bust_count / total
    p_safe = safe_count / total

    # 安全数字命中后期望加分
    if safe_count > 0:
        avg_safe_value = sum(v * c for v, c in counts.numbers.items() if v not in hand_set) / safe_count
    else:
        avg_safe_value = 0

    # 6 翻概率（仅当当前 5 张不同）
    p_six = p_safe if n_unique == 5 else 0
    p_continue_with_safe = p_safe if n_unique < 5 else 0

    p_flat = counts.bonus_flat / total
    p_double = counts.bonus_double / total
    p_ins = counts.insurance / total
    p_exile = counts.exile / total
    p_triple = counts.triple / total

    # EV 估算
    ev_fold = float(cur_score)
    if me.has_insurance:
        # 爆牌不归零（消耗保险）
        ev_bust_cost = 0  # 视为不变
    else:
        ev_bust_cost = -cur_score

    ev_draw = 0.0
    ev_draw += p_bust * (cur_score + ev_bust_cost)  # 爆牌路径
    ev_draw += p_continue_with_safe * (cur_score + avg_safe_value)  # 安全（未触六翻）
    ev_draw += p_six * (cur_score + avg_safe_value + 15)  # 6 翻终结
    ev_draw += p_flat * (cur_score + 10)
    ev_draw += p_double * (cur_score + sum(me.hand_numbers))  # 翻倍数字总和
    ev_draw += (p_ins + p_exile + p_triple) * (cur_score + 5)  # 技能粗估 +5

    # 调 expectimax 拿"最优应对后的 EV"
    expectimax = ExpectimaxAgent(depth=3)
    expectimax_decision = expectimax.choose_action(state, my_idx)

    return {
        "total_remaining": total,
        "n_unique": n_unique,
        "cur_score": cur_score,
        "p_bust": p_bust * 100,
        "p_safe": p_safe * 100,
        "p_six": p_six * 100,
        "avg_safe_value": avg_safe_value,
        "p_flat": p_flat * 100,
        "p_double": p_double * 100,
        "p_ins": p_ins * 100,
        "p_exile": p_exile * 100,
        "p_triple": p_triple * 100,
        "ev_draw": ev_draw,
        "ev_fold": ev_fold,
        "expectimax_decision": expectimax_decision,
    }


def render_analysis(a: dict, state: GameState, my_idx: int) -> None:
    if "error" in a:
        print(RED(f"\n  ⚠ {a['error']}"))
        return
    me = state.players[my_idx]
    target = state.config.target_score
    print()
    print(BOLD(CYAN("━━━━━━━━━━━━━━ 分析结果 ━━━━━━━━━━━━━━")))
    print(f"  剩余牌库：{a['total_remaining']} 张")
    print(f"  你的手牌：{sorted(me.hand_numbers)}（{a['n_unique']} 个不同数字）")
    print(f"  本局已得：{a['cur_score']} 分    保险：{'有' if me.has_insurance else '无'}")
    print(f"  你的总分：{me.total_score} / {target}")
    print()

    # 概率分解
    print(BOLD("  📊 抽下一张的概率分解："))
    bust_color = GREEN if a["p_bust"] < 15 else YELLOW if a["p_bust"] < 30 else RED
    bust_pct = bust_color(f"{a['p_bust']:>5.1f}%")
    bust_note = "有保险，免一次" if me.has_insurance else f"本局 {a['cur_score']} 分归零"
    print(f"    💥 爆牌：     {bust_pct}    （{bust_note}）")
    if a["n_unique"] < 5:
        safe_pct = GREEN(f"{a['p_safe']:>5.1f}%")
        print(f"    ✅ 安全数字： {safe_pct}    平均 +{a['avg_safe_value']:.1f} 分")
    elif a["n_unique"] == 5:
        six_pct_str = f"{a['p_six']:>5.1f}%"
        six_pct = BOLD(YELLOW(six_pct_str)) if a["p_six"] > 50 else YELLOW(six_pct_str)
        print(f"    🚀 6 翻终结：{six_pct}    +15 奖励 + 强制全场结算")
    print(f"    🟢 +10 加分： {a['p_flat']:>5.1f}%")
    print(f"    🟢 翻倍：     {a['p_double']:>5.1f}%    +{sum(me.hand_numbers)} 分")
    print(f"    🛡 保险：     {a['p_ins']:>5.1f}%")
    print(f"    🚷 放逐：     {a['p_exile']:>5.1f}%")
    print(f"    ⚡ 三连：     {a['p_triple']:>5.1f}%")
    print()

    # EV
    print(BOLD("  📈 期望值对比："))
    diff = a["ev_draw"] - a["ev_fold"]
    draw_label = f"EV(继续摸) = {a['ev_draw']:.1f}"
    fold_label = f"EV(跑路) = {a['ev_fold']:.1f}"
    if diff > 1:
        print(f"    {GREEN(draw_label)}   {DIM(fold_label)}")
        print(f"    差值：{GREEN(f'+{diff:.1f}')}")
    elif diff < -1:
        print(f"    {DIM(draw_label)}   {GREEN(fold_label)}")
        print(f"    差值：{RED(f'{diff:.1f}')}")
    else:
        print(f"    {YELLOW(draw_label)}   {YELLOW(fold_label)}")
        print(f"    差值：{YELLOW(f'{diff:+.1f}')}（接近，看风险偏好）")
    print()

    # 推荐
    decision = a["expectimax_decision"]
    if decision == "draw":
        rec = BOLD(GREEN("🎲 继续摸牌"))
    else:
        rec = BOLD(YELLOW("🔒 跑路（锁分）"))
    print(BOLD(f"  💡 Expectimax 建议：{rec}"))

    # 特殊提示
    if me.total_score + a["cur_score"] >= target:
        print(BOLD(GREEN(f"  🏆 警告：现在跑路即可获胜！锁定 {a['cur_score']} 分赢比赛。")))
    if a["n_unique"] == 5 and a["p_six"] > 50:
        print(BOLD(YELLOW(f"  🚀 5 张冲刺机会：6 翻概率 {a['p_six']:.1f}%，建议拼一把")))
    if a["p_bust"] > 35:
        print(BOLD(RED(f"  ⚠ 高风险：爆牌概率 {a['p_bust']:.1f}%，强烈建议跑路")))


def interactive() -> None:
    print(BOLD(CYAN("━━━━━━━━━━━━━━ 决策助手 ━━━━━━━━━━━━━━")))
    print(DIM("  输入当前局面，按回车跳过的字段使用默认值\n"))

    hand_str = input("  我的手牌（数字，空格分隔，如 '3 5 8'）: ").strip()
    my_hand = parse_hand(hand_str)

    bonus_str = input("  我的加分总分（默认 0）: ").strip() or "0"
    my_bonus = int(bonus_str)
    # 加分牌张数：用户给加分总分，从 +2/+4/+6/+8/+10 推算最少需要几张能凑出
    # （advisor 只关心剩余牌库，所以用合理估算即可）
    my_bonus_flat_count = 0
    remaining_bonus = my_bonus
    for v in [10, 8, 6, 4, 2]:
        if remaining_bonus >= v:
            my_bonus_flat_count += 1
            remaining_bonus -= v
    my_bonus_double_count = 0
    if input("  抽过翻倍牌？(y/n) [n]: ").strip().lower() in ("y", "yes"):
        my_bonus_double_count = 1
    my_insurance = input("  有保险？(y/n) [n]: ").strip().lower() in ("y", "yes")

    total_str = input("  我的总分（默认 0）: ").strip() or "0"
    my_total = int(total_str)

    target_str = input("  目标分（默认 200）: ").strip() or "200"
    target = int(target_str)

    n_str = input("  总玩家数（默认 4）: ").strip() or "4"
    n_players = int(n_str)

    # 简化版：不细问对手手牌，只问已见牌总览
    print(DIM("\n  已见过的其他牌（含其他玩家手牌+弃牌堆，可不输）"))
    print(DIM("  格式示例：'1 1 5 7 +10 保险 翻倍'"))
    seen_str = input("  > ").strip()
    seen_nums, seen_specials = parse_seen(seen_str)

    # 构建状态
    rem = build_remaining(my_hand, my_bonus_flat_count, my_bonus_double_count,
                           my_insurance, seen_nums, seen_specials)

    cfg = GameConfig(num_players=n_players, target_score=target)
    me = PlayerState(index=0, total_score=my_total)
    me.hand_numbers = list(my_hand)
    me.bonus_flat_total = my_bonus
    me.has_insurance = my_insurance
    me.status = PlayerStatus.ACTIVE
    others = [PlayerState(index=i, total_score=0, status=PlayerStatus.ACTIVE)
              for i in range(1, n_players)]
    state = GameState(config=cfg, players=[me] + others, remaining=rem)

    a = analyze(state, my_idx=0)
    render_analysis(a, state, my_idx=0)


def main() -> None:
    parser = argparse.ArgumentParser(description="决策助手：建议 draw 或 fold")
    parser.add_argument("--hand", help="我的手牌，如 '3 5 8'")
    parser.add_argument("--bonus", type=int, default=0, help="我的加分总分")
    parser.add_argument("--double", action="store_true", help="抽过翻倍牌")
    parser.add_argument("--insurance", action="store_true", help="有保险")
    parser.add_argument("--total", type=int, default=0, help="我的总分")
    parser.add_argument("--target", type=int, default=200, help="目标分")
    parser.add_argument("--players", type=int, default=4, help="玩家数")
    parser.add_argument("--seen", default="", help="已见过的牌，如 '1 5 +10 保险'")
    args = parser.parse_args()

    if args.hand is None:
        interactive()
        return

    my_hand = parse_hand(args.hand)
    # 同 interactive 推算 flat 张数
    my_bonus_flat_count = 0
    remaining = args.bonus
    for v in [10, 8, 6, 4, 2]:
        if remaining >= v:
            my_bonus_flat_count += 1
            remaining -= v
    my_bonus_double_count = 1 if args.double else 0
    seen_nums, seen_specials = parse_seen(args.seen)
    rem = build_remaining(my_hand, my_bonus_flat_count, my_bonus_double_count,
                           args.insurance, seen_nums, seen_specials)
    cfg = GameConfig(num_players=args.players, target_score=args.target)
    me = PlayerState(index=0, total_score=args.total)
    me.hand_numbers = list(my_hand)
    me.bonus_flat_total = args.bonus
    me.has_insurance = args.insurance
    me.status = PlayerStatus.ACTIVE
    others = [PlayerState(index=i, total_score=0, status=PlayerStatus.ACTIVE)
              for i in range(1, args.players)]
    state = GameState(config=cfg, players=[me] + others, remaining=rem)
    a = analyze(state, my_idx=0)
    render_analysis(a, state, my_idx=0)


if __name__ == "__main__":
    main()
